python/lsst/analysis/ap/apdb.py

```diff
@@ -0,0 +1,243 @@
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""APDB connection management and data access tools.
+"""
+
+__all__ = ["DbQuery", "ApdbSqliteQuery", "ApdbPostgresQuery"]
+
+import abc
+import contextlib
+import sqlite3
+import warnings
+
+import pandas as pd
+import psycopg
+
+
+class DbQuery(abc.ABC):
+    """Base class for APDB connection and query management.
+
+    Subclasses must specify a ``connection`` property to use as a context-
+    manager for queries.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler to unpack detector/visit from ccdVisitId.
+        To be deprecated once this information is in the database.
+    instrument : `str`
+        Short name (e.g. "DECam") of instrument to make a dataId unpacker
+        and to add to the table columns; supports any gen3 instrument.
+        To be deprecated once this information is in the database.
+    verbose : `bool`
+        Be verbose when sending commands, e.g. print each query as it is run.
+    """
+    def __init__(self, *, butler, instrument, verbose=False):
+        self._butler = butler
+        self._instrument = instrument
+        self._verbose = verbose
+
+    @property
+    @contextlib.contextmanager
+    @abc.abstractmethod
+    def connection(self):
+        """Context manager for database connections.
+
+        Must yield an sql Connection object like ``psycopg.Connection``;
+        whether the connection is closed after the context manager is closed
+        is implementation dependent.
+        """
+        pass
+
+    def load_sources_for_object(self, dia_object_id, limit=100000):
+        """Load diaSources for a single diaObject.
+
+        Parameters
+        ----------
+        dia_object_id : `int`
+            Id of object to load sources for.
+        limit : `int`
+            Maximum number of rows to return.
+
+        Returns
+        -------
+        data : `pandas.DataFrame`
+            DiaSources for the specified diaObject.
+        """
+        with self.connection as connection:
+            order = 'ORDER BY "ccdVisitId", "diaSourceId"'
```

Comment:
Yes, any of these table or column names that I've double-quoted. This is the danger of writing sql code that you expect to work on different sql implementations (postgres vs. sqlite here).

I've since learned that it's a known misfeature in the sqlite spec.

https://www.sqlite.org/quirks.html#double_quoted_string_literals_are_accepted

Reply:
Nice